### DISCUSSION TITLE
Migration to newest java version

### DISCUSSION THREAD

**Discussion by giusvale-dev** [jorabin/KeePassJava2#46] [Ideas]
Hi @jorabin,

First of all, this is great work! Congratulations!

It's very interesting and can be useful in many contexts by incorporating this API into other projects. However, I've noticed that the Java version being used is quite old, and this could potentially pose a problem if you plan to use a newer version of Java.

Do you think it would be possible to migrate to Java version 20?

Giuseppe

**Comment by jorabin** [comment]
Hi Giuseppe

I think that you could recompile for Java 20 but be aware that the SimpleXML implementation probably won't work beyond Java 17 - see https://github.com/jorabin/KeePassJava2/issues/40 but I would think would be OK for JAXB and DOM. Would be good to hear how you get on.

Jo

**Reply by giusvale-dev** [reply]
Hi Jo,
I tested java 20 with this project, and I confirm that "simple" module doesn't work (test fails) as you reported!

I think it could be a good idea to migrate from SimpleXML to JAXB directly (but I don't know if you want to continue to use JAXB), and then you should remove SimpleXML from the project. (This is my current try).

I personally don't like the XML as complex data storage, I prefer RDBMS (SQLite in this case) but I understand that this is a hard task (maybe impossible to let compatibility with others KeePass applications).

However, I'd like to contribute to this project.

When you have time, give me a feedback, I'd like to hear your way.

Giuseppe

**Comment by jorabin** [comment]
Hi again Giuseppe

KeePass is an encrypted XML file format, so there is no getting away from the need to serialise and deserialise XML to and from KeePass conformant files. The various current implementations here store as an in-memory XML DOM (DomDatabaseWrapper) or as Java Objects. Of course you can write an adapter that writes to a database. That's the point of having built the whole thing this way.

As for how to move on from SimpleXML, question is posed at #40 ... I think Jackson the top candidate.

**Comment by giusvale-dev** [comment]
Dear Jo,
Thanks a lot for your explanation.
Now I understand the project architecture and see the block diagram, which is now clear to 100%.
I apologize for the previous confusion. 

I'd like to integrate the new jackson module on the project, and I'd like to send you a pull request as soon as possible.

I'll test this maven module with java 8.

For java version 20, I think it is needed to remove the simple module from the project and test all modules again (I will try also with java 20).

Giuseppe

**Comment by jorabin** [comment]
Hopefully you'll have seen that the existing implementations conform to the same code structure illustrated [here](https://github.com/jorabin/KeePassJava2/blob/master/DatabaseStructure.svg) and unless there's a good reason not to would be good if any new implementations conform to that pattern.

Also please make pull requests to develop.

Thanks!

**Comment by giusvale-dev** [comment]
Hi Jo,
I'm developing the Jackson module on the KeePassJava project.
All tests now are passing except for 2 described above:

test123 The decrypt here fails, I need to investigate again because I don't understand why this fails (the key is the same, the ciphertext and IV are equals, for now, I don't know why this not working).

upperCaseBooleanTest
In this case, the module writes the attribute Protected in all cases (true and false) and this isn't the correct way.

The good news is that most part of the Jackson module is ready, 
I need to solve these 2 issues before sending you the pull request.

I will update you ASAP

Giuseppe



![image](https://github.com/jorabin/KeePassJava2/assets/6003216/89f59a85-ce6f-42ee-ae3c-404dfe392063)

**Reply by jorabin** [reply]
Are you decrypting the protected fields on load? Note that protected fields are additionally encrypted (additional to the encryption of the file) using the "inner stream cipher" - one of the principal jobs of the SerializableDatabase implementations is to encrypt and decrypt those fields on load and save.

I'm curious to see where you have got to on this, will you push your implementation to your forked repo? Presumably [this one](https://github.com/giusvale-dev/KeePassJava2)

**Reply by giusvale-dev** [reply]
All problems are fixed, jackson now is a new module of the project!
Solved in deserialization/serialization phases

**Reply by jorabin** [reply]
well done!

I asked for a couple of tweaks before merging, thanks.

**Comment by jorabin** [comment]
Well, that is now on main as 2.2.2-SNAPSHOT ... good work, Giuseppe!

**Comment by jorabin** [comment]
A couple of "quirks" #50 and #51 come out of a bit of testing. But on the whole this is looking solid! Worth noting that it's not completely straightforward to stop Jackson serializing everything in sight, including "getters" in superclasses, which you would not want to annotate with things that are only relevant to e.g. the Jackson subclass. So if you want to serialize only the annotate fields then something like [this](https://github.com/jorabin/KeePassJava2/blob/4a007f3e6cd14a2d826e719210c77f7a2de8cf7f/jackson/src/main/java/org/linguafranca/pwdb/kdbx/jackson/JacksonSerializableDatabase.java#L92) is your answer.

**Reply by giusvale-dev** [reply]
Dear Jo, Thanks for your speedy corrections!

**Comment by jorabin** [comment]
For version 3 I think we'll go with Java 11 as the base line. 

Comments welcome under #62 

Many thanks
